lib/python/rosie/ws.py

```diff
@@ -19,77 +19,158 @@
+        self.stopping = False
+
         self.props = {}
         rose_conf = ResourceLocator.default().get_conf()
         self.props["title"] = rose_conf.get_value(
             ["rosie-disco", "title"], self.TITLE)
         self.props["host_name"] = rose_conf.get_value(["rosie-disco", "host"])
         if self.props["host_name"] is None:
-            self.props["host_name"] = HostSelector().get_local_host()
+            self.props["host_name"] = socket.gethostname()
             if self.props["host_name"] and "." in self.props["host_name"]:
                 self.props["host_name"] = (
                     self.props["host_name"].split(".", 1)[0])
         self.props["rose_version"] = ResourceLocator.default().get_version()
+        # Autoescape markup to prevent code injection from user inputs.
         self.props["template_env"] = jinja2.Environment(
             loader=jinja2.FileSystemLoader(
                 ResourceLocator.default().get_util_home(
-                    "lib", "html", "template", "rosie-disco")))
+                    "lib", "html", "template", "rosie-disco")),
+            autoescape=jinja2.select_autoescape(
+                enabled_extensions=('html', 'xml'), default_for_string=True))
+
         db_url_map = {}
-        for key, node in rose_conf.get(["rosie-db"]).value.items():
+        for key, node in list(rose_conf.get(["rosie-db"]).value.items()):
             if key.startswith("db.") and key[3:]:
                 db_url_map[key[3:]] = node.value
         self.db_url_map = db_url_map
-        if not self.db_url_map:
-            self.db_url_map = {}
-        for key, db_url in self.db_url_map.items():
-            setattr(self, key, RosieDiscoService(self.props, key, db_url))
 
-    @cherrypy.expose
-    def index(self, *_):
+        # Set-up the Tornado application request-handling structure.
+        service_root = r"/" + self.NAMESPACE + r"/?"
+        prefix_handlers = []
+        class_args = {"props": self.props}
+
+        root_class_args = dict(class_args)  # mutable so copy for safety
+        root_class_args.update({"db_url_map": self.db_url_map})
+        root_handler = (service_root, RosieDiscoServiceRoot,
+                        root_class_args)
+        for key, db_url in list(self.db_url_map.items()):
+            prefix_class_args = dict(class_args)  # mutable so copy for safety
+            prefix_class_args.update({
+                "prefix": key,
+                "db_url": db_url,
+                "root_service_url": self.NAMESPACE,
+            })
+            handler = (service_root + key + r"/?$", RosieDiscoService,
+                       prefix_class_args)
+            hello_handler = (service_root + key + r"/hello/?$",
+                             HelloHandler, prefix_class_args)
+            search_handler = (service_root + key + r"/search(.*)",
+                              SearchHandler, prefix_class_args)
+            query_handler = (service_root + key + r"/query(.*)",
```

Comment:
That sounds like a more elegant approach. I'll have a play around to see if I can get it working. Thanks.

Reply:
[See point 4 of https://github.com/metomi/rose/pull/2288#issuecomment-476096016 (which upon merge will be moved to an Issue): this will be follow-on work to this PR.]